client/dom/use-context-menu.tsx

```diff
@@ -0,0 +1,73 @@
+import React, { useCallback, useState } from 'react'
+import { OriginX, OriginY, PopoverProps, usePopoverController } from '../material/popover'
+
+interface UseContextMenuProps {
+  originX?: OriginX
+  originY?: OriginY
+}
+
+/**
+ * React hook that can be used to setup context menu on any element.
+ *
+ * @example
+ *
+ * const Component = () => {
+ *   const { onContextMenu, contextMenuPopoverProps} = useContextMenu()
+ *   return (
+ *     <>
+ *       <Popover {...contextMenuPopoverProps}>
+ *         <span>Hi, I'm in context menu!</span>
+ *       </Popover>
+ *       <span onContextMenu={onContextMenu}>Hello world!</span>
+ *     </>
+ *   )
+ * }
+ */
+export function useContextMenu(props?: UseContextMenuProps): {
+  onContextMenu: (event: React.MouseEvent) => void
+  contextMenuPopoverProps: Omit<PopoverProps, 'children'>
+} {
+  const [anchor, setAnchor] = useState<Element>()
+  const [anchorX, setAnchorX] = useState(0)
+  const [anchorY, setAnchorY] = useState(0)
+
+  const [isContextMenuOpen, openContextMenu, closeContextMenu] = usePopoverController()
+
+  const onOpen = useCallback(
+    (event: React.MouseEvent) => {
+      event.preventDefault()
+
+      setAnchorX(event.pageX)
+      setAnchorY(event.pageY)
+      setAnchor(event.currentTarget)
+      openContextMenu(event)
+    },
+    [openContextMenu],
+  )
+  const onDismiss = useCallback(
+    (event?: MouseEvent) => {
+      if (
+        !anchor ||
+        event?.button !== 2 ||
+        !event?.target ||
+        !anchor.contains(event.target as Node)
+      ) {
+        setAnchor(undefined)
+        closeContextMenu()
+      }
```

Comment:
Not sure I fully understand what you mean.

Reply:
If they are right-clicking inside of the anchor, the expected behavior is that the popover will move to the new point they right clicked. Your implementation simply leaves the popover where it was

Reply:
It's possible it works via the `onOpen` handler firing again but:

A) That should probably be noted here, as it's non-obvious
B) You should definitely know that for a fact before merging this code, just because code is pulled from elsewhere doesn't mean it can be merged without testing it. Please at least do a check of all of the places you've modified in this PR